AEP/aep-00007.md

```diff
@@ -0,0 +1,225 @@
+This design allows to have a task ready for execution as soon as the previous task ends.
+Tasks undergoing retry are treated as entirely new tasks, which simplifies tracking and execution, generating new messages accordingly.
+
+Message uniqueness is not required as it is managed elsewhere.
+Furthermore, message scheduling must be handled within the queue service to accommodate prioritization mechanisms.
+The system should also allow for the seamless integration of new plugins to enhance flexibility and adaptability.
+
+## Possible Approaches for the Interface
+
+Several approaches can be considered for implementing the message processing interface, including event-driven and pull-based mechanisms. Given the need to control pipelining—managing the number of concurrently processed messages and determining when processing begins—both approaches are functionally equivalent. In practice, polling would be used on an internal queue to regulate message processing.
+
+Below is an overview of how major market solutions implement message retrieval mechanisms:
+
+| Solution      | Approche |
+| ------------- | -------- |
+| AWS SQS       | pull     |
+| ActiveMQ      | both     |
+| RabbitMQ      | both     |
+| Google PubSub | both     |
+| Pulsar        | both     |
+
+## Trade-offs and Constraints
+
+Rewriting the codebase to adopt a fully event-driven, push-based pattern is not feasible given the complexity and scale of such a transformation. The current design prioritizes simplicity and robustness over a pure event-driven model. Considering the long duration of tasks and the cost of orchestration, some inefficiencies during message reception are acceptable. These inefficiencies are preferable to the development overhead and technical debt involved in transforming the existing pull-based mechanism.
+
+The technical limitations of the underlying queueing service also play a role in the design decisions. SQS, for instance, does not provide an API for push-based message reception, which necessitates the use of a pull-based approach. Additionally, past attempts to use RabbitMQ for push-based message handling revealed significant stability issues, such as connection losses and inconsistent message processing. These challenges have further reinforced the decision to rely on the pull-based model, which offers greater reliability and predictability.
+
+We can consider converting a push-based reception to messages into our pull-based interfaces. However, this implementation requires to be as stable as the current one. Pull requests are welcome to improve the implementation of our queue system adaptors.
+
+## Current Interface
+
+Queue interfaces can be found in our dotnet package [ArmoniK.Core](https://www.nuget.org/packages/ArmoniK.Core.Base).
+It contains everything needed to implement queue plugins that can be dynamically loaded by ArmoniK.Core allowing users to implement plugins that match their requirements.
```

Comment:
I'd write "The package provides the elements necessary to implement queue plugins ... "